azure/services/managedclusters/managedclusters.go

```diff
@@ -65,22 +68,43 @@ type ManagedClusterScope interface {
 	SetAutoUpgradeVersionStatus(version string)
 	SetVersionStatus(version string)
 	IsManagedVersionUpgrade() bool
+	IsPreviewEnabled() bool
 }
 
 // New creates a new service.
-func New(scope ManagedClusterScope) *aso.Service[*asocontainerservicev1.ManagedCluster, ManagedClusterScope] {
-	svc := aso.NewService[*asocontainerservicev1.ManagedCluster](serviceName, scope)
-	svc.Specs = []azure.ASOResourceSpecGetter[*asocontainerservicev1.ManagedCluster]{scope.ManagedClusterSpec()}
+func New(scope ManagedClusterScope) *aso.Service[genruntime.MetaObject, ManagedClusterScope] {
+	// genruntime.MetaObject is used here instead of an *asocontainerservicev1.ManagedCluster to better
+	// facilitate returning different API versions.
+	svc := aso.NewService[genruntime.MetaObject](serviceName, scope)
+	svc.Specs = []azure.ASOResourceSpecGetter[genruntime.MetaObject]{scope.ManagedClusterSpec()}
 	svc.ConditionType = infrav1.ManagedClusterRunningCondition
 	svc.PostCreateOrUpdateResourceHook = postCreateOrUpdateResourceHook
 	return svc
 }
 
-func postCreateOrUpdateResourceHook(ctx context.Context, scope ManagedClusterScope, managedCluster *asocontainerservicev1.ManagedCluster, err error) error {
+func postCreateOrUpdateResourceHook(ctx context.Context, scope ManagedClusterScope, obj genruntime.MetaObject, err error) error {
 	if err != nil {
 		return err
 	}
 
+	// If existing is preview, convert to stable for this function.
+	var existing *asocontainerservicev1.ManagedCluster
+	if scope.IsPreviewEnabled() {
+		existingPreview := obj.(*asocontainerservicev1preview.ManagedCluster)
+		hub := &asocontainerservicev1hub.ManagedCluster{}
+		if err := existingPreview.ConvertTo(hub); err != nil {
+			return err
+		}
+		prev := &asocontainerservicev1.ManagedCluster{}
+		if err := prev.ConvertFrom(hub); err != nil {
```

Comment:
Like in `Parameters`, maybe we don't need to convert to any particular version again and we can stick with using the hub type here.

Reply:
Do we wanna make this change before merging?

Reply:
I wouldn't block on this.